### PR TITLE
User login errors

### PR DESCRIPTION
**Pull request by ThomasMcDonnell**
First off congrats, this is great and you've done an awesome job 👍 I noticed a couple of things and thought this would be a great moment to help you on the way. One of the things I love the most about working in software is the code reviews from my colleagues who know things I don't and have been doing this longer than me, it is the single best way to learn new things and new ways of doing things 😄 So in the spirit of bouncing around ideas and helping out in a small way I have PR'd a fix for the login error handling and added a couple of pointers regarding structure ect.

Feel free to ignore the PR if you disagree with anything or reach out if you have questions, hope this helps you either way 😄 

**Changes log**

- login errors before

image
image

- login errors after

image
image

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -14,15 +14,15 @@ app.use(bodyParser.urlencoded({ extended: true }))
 app.set('trust proxy', 1)
 app.use(cookieSession({
     name: 'session',
-    keys: [`${process.env.SECRET_1}`, `${process.env.SECRET_2}`]
+    keys: ['secret1', "secrete2"]
```
Comment (ThomasMcDonnell) [inline app.js:17]:
```suggestion
    keys: [`${process.env.SECRET_1}`, `${process.env.SECRET_2}`]
```

File: routes/routes.js [hunk 2]
```diff
@@ -4,7 +4,9 @@ User = require('../models/user')
 Roster = require('../models/roster')
 const { check, validationResult} = require("express-validator");
 const bcrypt = require("bcryptjs");
-const config = require('../config');
+// const config = require('../config');
```
Comment (ThomasMcDonnell) [inline routes/routes.js:7]:
```suggestion
const config = require('../config');
```

File: views/rosterview/login.hbs [hunk 3]
```diff
@@ -2,13 +2,22 @@
...
 <form action="/user/login" method="post">
     <div class="col s12">
-        <input type="email" name="email" class="text col s4 offset-m4" placeholder="Email">
-        <input type="password" name="password" class="text col s4 offset-m4" placeholder="Password">
-        <div class="col s3 offset-m6">
-            <button class="btn waves-effect waves-light col s7" type="submit">Login
-                <i class="material-icons right">login</i/>
-            </button>
-        </div>
+        <input type="email" name="email" class="text col s4 offset-m4" placeholder="Email" required>
```
Comment (ThomasMcDonnell) [inline views/rosterview/login.hbs:5]: You are using the browser to post back to the server, you can just add the required flag and push validation on to the browser when submitting. No need then to validate on the server if this is a valid email format. Similar situation with the password, browser wont submit unless present with a valid email.

File: views/rosterview/login.hbs [hunk 4]
```diff
@@ -2,13 +2,22 @@
...
     </div>
 </form>
 
+{{#if errors }}
+<script>
+    var error = "Opps, {{errors}}"
+    document.addEventListener('DOMContentLoaded', (event) => {
+        M.toast({html: error})
+    });
+</script>
```
Comment (ThomasMcDonnell) [inline views/rosterview/login.hbs:21]: There are lots of ways to do this but seemed easy just to add the script to the login and let handlebars decide if to have it execute or not dependent on the presence of the errors.

File: app.js [hunk 5]
```diff
@@ -21,8 +21,7 @@ app.set('view engine', 'hbs');
 app.engine('hbs', exphbs({extname: 'hbs', defaultLayout: 'mainLayout', layoutsDir: __dirname + '/views/layouts/'}))
 
 
-const rosterRoute = require('./routes/routes')
-app.use('/', rosterRoute)
-
-app.listen(port, () => console.log(`Roster App listening on port ` + port))
+const rosterRoute = require('./routes/routes');
+app.use('/', rosterRoute);
```
Comment (ThomasMcDonnell) [inline app.js:25]: Added the` ; `because I am one of those 😂

File: db/db.js [hunk 6]
```diff
@@ -1,11 +1,15 @@
...
-    //Export module
-    module.exports = mongoose;
\ No newline at end of file
+const uri = `${process.env.MONGODB_URL}`;
+mongoose.connect(uri, {
+    useNewUrlParser: true,
+    useCreateIndex: true,
+    useFindAndModify: false,
+    useUnifiedTopology: true
+});
```
Comment (ThomasMcDonnell) [inline db/db.js:11]: Just some extra defaults that are handy to have with mongo

File: routes/routes.js [hunk 7]
```diff
@@ -195,54 +197,6 @@ router.post(
...
-      
-    } catch (e) {
-      console.error(e);
-      res.status(500).json({
-        message: "Server Error"
-      });
-    }
-  }
-);
+router.post("/user/login", authController.login);
```
Comment (ThomasMcDonnell) [inline routes/routes.js:200]: A common pattern you will see is MVC or model view controller, most larger frameworks that favour convention over configuration will follow this pattern eg Rails, Laravel, Django (well Django is technically model view template 🤷 ). Express is very unopinionated in this regard but its a great habit to get into, I think it makes things cleaner and the code more organised. Helps when navigating a larger code base.

So here you can see clearly the route and the controller method thats passed in as a handler.

File: controllers/authController.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+      user_id: user.id
+    });
+  } else {
+    req.session.logged = false;
+    res.render("rosterview/login", {
+      viewTitle: "Login",
+      errors: "Email & Password don't match!"
+    });
+  }
+}
```
Comment (ThomasMcDonnell) [inline controllers/authController.js:24]: So here is the corresponding controller for auth and the login method. A couple of changes here first being re-rendering the login template on unauthorized with an extra context variable or errors. There are lots of ways you can do this before you were rendering a JSON error response, which is fine if you're handling the login through ajax. But seeing as you were using handlebars and submitting via the browser seems reasonable and simpler just to let handlebars do the heavy lifting and re-render the login page with errors present.

One thing to note I haven't handled the async error, you can do as before in a try-catch and render a 500 error template page, or you can wrap the login method in a function like `catchAsyncError(fn) => { fn(req, res, next).catch(next) }` and then handle as you see fit.

File: routes/routes.js [hunk 9]
```diff
@@ -6,6 +6,8 @@ const { check, validationResult} = require("express-validator");
 const bcrypt = require("bcryptjs");
 const config = require('../config');
 
+const authController = require('../controllers/authController');
```
Comment (ThomasMcDonnell) [inline routes/routes.js:9]: You could imagine this file would become very condensed down, with just loading the controllers in and defining the routes passing the handlers in.

File: routes/routes.js (hunk 7 above)
Comment (ThomasMcDonnell) [reply routes/routes.js:200]: Also no need to validate the email here anymore or the password, the browser will do that for you and validate the length of the password on login as its already been created 😄

**Comment by RaphaelQSmith** [comment]
Thanks for the PR, Thomas. Very insightful and detailed suggestions that I'll definitely implement on my projects from now on.

**Pull request merged**